In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import svm
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import HistGradientBoostingClassifier
from lightgbm import LGBMClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import mean_absolute_error, mean_absolute_percentage_error, r2_score, mean_squared_error, accuracy_score

In [2]:
#importing Dataset
data = pd.read_csv("asdsd.csv")
data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [3]:
data.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [4]:
data['Tenure'] = pd.cut(data['tenure'],
                      bins=[0, 12, 24, 36, 48, 60, np.inf],
                      labels=[0, 1, 2, 3, 4, 5])

In [5]:
data.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... TechSupport  \
0  No phone service             DSL             No  ...          No   
1                No             DSL            Yes  ...          No   
2                No             DSL            Yes  ...          No   
3  No phone service             DSL            Yes  ...         Yes   
4                No     Fiber optic             No  ...          No   

  StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No              No  Month-to-month              Yes   
1          No              No        One year               No   
2          No              No  Month-to-month              Yes   
3          No              No        One year               No   
4          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges TotalCharges  Churn Tenure  
0           Electronic check          29.85        29.85     No      0  
1               Mailed check          56.95       1889.5     No      2  
2               Mailed check          53.85       108.15    Yes      0  
3  Bank transfer (automatic)          42.30      1840.75     No      3  
4           Electronic check          70.70       151.65    Yes      0  

[5 rows x 22 columns]

In [6]:
#'Female - 0', 'Male - 1'
data.gender.unique()

array(['Female', 'Male'], dtype=object)

In [7]:
data.SeniorCitizen.unique()

array([0, 1], dtype=int64)

In [8]:
#'Yes - 0', 'No - 1'
data.Partner.unique()

array(['Yes', 'No'], dtype=object)

In [9]:
#'No - 0', 'Yes - 1'
data.Dependents.unique()

array(['No', 'Yes'], dtype=object)

In [10]:
#'No - 0', 'Yes - 1'
data.PhoneService.unique()

array(['No', 'Yes'], dtype=object)

In [11]:
#'No phone service - 0', 'No - 1', 'Yes - 2'
data.MultipleLines.unique()

array(['No phone service', 'No', 'Yes'], dtype=object)

In [12]:
#'DSL - 0', 'Fiber optic - 1', 'No - 2'
data.InternetService.unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [13]:
#'No - 0', 'Yes - 1', 'No internet service - 2'
data.OnlineSecurity.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [14]:
#'Yes - 0', 'No - 1', 'No internet service - 2'
data.OnlineBackup.unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [15]:
#'No - 0', 'Yes - 1', 'No internet service - 2'
data.DeviceProtection.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [16]:
#'No - 0', 'Yes - 1', 'No internet service - 2'
data.TechSupport.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [17]:
#'No - 0', 'Yes - 1', 'No internet service - 2'
data.StreamingTV.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [18]:
#'No - 0', 'Yes - 1', 'No internet service - 2'
data.StreamingMovies.unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [19]:
#'Month-to-month - 0', 'One year - 1', 'Two year - 2'
data.Contract.unique()

array(['Month-to-month', 'One year', 'Two year'], dtype=object)

In [20]:
#'Yes - 0', 'No - 1'
data.PaperlessBilling.unique()

array(['Yes', 'No'], dtype=object)

In [21]:
#'Electronic check - 0 ', 'Mailed check - 1', 'Bank transfer (automatic) - 2','Credit card (automatic) - 3'
data.PaymentMethod.unique()

array(['Electronic check', 'Mailed check', 'Bank transfer (automatic)',
       'Credit card (automatic)'], dtype=object)

In [22]:
#No - 0; Yes - 1
data.Churn.unique()

array(['No', 'Yes'], dtype=object)

In [23]:
le = LabelEncoder()
data['gender'] = le.fit_transform(data.gender.values)
data['SeniorCitizen'] = le.fit_transform(data.SeniorCitizen.values)
data['Partner'] = le.fit_transform(data.Partner.values)
data['Dependents'] = le.fit_transform(data.Dependents.values)
data['PhoneService'] = le.fit_transform(data.PhoneService.values)
data['MultipleLines'] = le.fit_transform(data.MultipleLines.values)
data['InternetService'] = le.fit_transform(data.InternetService.values)
data['OnlineSecurity'] = le.fit_transform(data.OnlineSecurity.values)
data['OnlineBackup'] = le.fit_transform(data.OnlineBackup.values)
data['DeviceProtection'] = le.fit_transform(data.DeviceProtection.values)
data['TechSupport'] = le.fit_transform(data.TechSupport.values)
data['StreamingTV'] = le.fit_transform(data.StreamingTV.values)
data['StreamingMovies'] = le.fit_transform(data.StreamingMovies.values)
data['Contract'] = le.fit_transform(data.Contract.values)
data['PaperlessBilling'] = le.fit_transform(data.PaperlessBilling.values)
data['PaymentMethod'] = le.fit_transform(data.PaymentMethod.values)

In [24]:
data['Churn'] = le.fit_transform(data.Churn.values)

In [25]:
data.Churn.unique()

array([0, 1])

In [26]:
data.Contract.unique()

array([0, 1, 2])

In [27]:
data = data.drop(['customerID','tenure'], axis=1)

In [28]:
data.head()

gender  SeniorCitizen  Partner  Dependents  PhoneService  MultipleLines  \
0       0              0        1           0             0              1   
1       1              0        0           0             1              0   
2       1              0        0           0             1              0   
3       1              0        0           0             0              1   
4       0              0        0           0             1              0   

   InternetService  OnlineSecurity  OnlineBackup  DeviceProtection  \
0                0               0             2                 0   
1                0               2             0                 2   
2                0               2             2                 0   
3                0               2             0                 2   
4                1               0             0                 0   

   TechSupport  StreamingTV  StreamingMovies  Contract  PaperlessBilling  \
0            0            0                0         0                 1   
1            0            0                0         1                 0   
2            0            0                0         0                 1   
3            2            0                0         1                 0   
4            0            0                0         0                 1   

   PaymentMethod  MonthlyCharges TotalCharges  Churn Tenure  
0              2           29.85        29.85      0      0  
1              3           56.95       1889.5      0      2  
2              3           53.85       108.15      1      0  
3              0           42.30      1840.75      0      3  
4              2           70.70       151.65      1      0

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   gender            7043 non-null   int32   
 1   SeniorCitizen     7043 non-null   int64   
 2   Partner           7043 non-null   int32   
 3   Dependents        7043 non-null   int32   
 4   PhoneService      7043 non-null   int32   
 5   MultipleLines     7043 non-null   int32   
 6   InternetService   7043 non-null   int32   
 7   OnlineSecurity    7043 non-null   int32   
 8   OnlineBackup      7043 non-null   int32   
 9   DeviceProtection  7043 non-null   int32   
 10  TechSupport       7043 non-null   int32   
 11  StreamingTV       7043 non-null   int32   
 12  StreamingMovies   7043 non-null   int32   
 13  Contract          7043 non-null   int32   
 14  PaperlessBilling  7043 non-null   int32   
 15  PaymentMethod     7043 non-null   int32   
 16  MonthlyCharges    7043 n

In [30]:
data['TotalCharges'] = pd.to_numeric(data['TotalCharges'], errors='coerce')

In [31]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   gender            7043 non-null   int32   
 1   SeniorCitizen     7043 non-null   int64   
 2   Partner           7043 non-null   int32   
 3   Dependents        7043 non-null   int32   
 4   PhoneService      7043 non-null   int32   
 5   MultipleLines     7043 non-null   int32   
 6   InternetService   7043 non-null   int32   
 7   OnlineSecurity    7043 non-null   int32   
 8   OnlineBackup      7043 non-null   int32   
 9   DeviceProtection  7043 non-null   int32   
 10  TechSupport       7043 non-null   int32   
 11  StreamingTV       7043 non-null   int32   
 12  StreamingMovies   7043 non-null   int32   
 13  Contract          7043 non-null   int32   
 14  PaperlessBilling  7043 non-null   int32   
 15  PaymentMethod     7043 non-null   int32   
 16  MonthlyCharges    7043 n

In [32]:
# Check the descriptive statistics of numeric variables
data.describe()

gender  SeniorCitizen      Partner   Dependents  PhoneService  \
count  7043.000000    7043.000000  7043.000000  7043.000000   7043.000000   
mean      0.504756       0.162147     0.483033     0.299588      0.903166   
std       0.500013       0.368612     0.499748     0.458110      0.295752   
min       0.000000       0.000000     0.000000     0.000000      0.000000   
25%       0.000000       0.000000     0.000000     0.000000      1.000000   
50%       1.000000       0.000000     0.000000     0.000000      1.000000   
75%       1.000000       0.000000     1.000000     1.000000      1.000000   
max       1.000000       1.000000     1.000000     1.000000      1.000000   

       MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
count    7043.000000      7043.000000     7043.000000   7043.000000   
mean        0.940508         0.872923        0.790004      0.906432   
std         0.948554         0.737796        0.859848      0.880162   
min         0.000000         0.000000        0.000000      0.000000   
25%         0.000000         0.000000        0.000000      0.000000   
50%         1.000000         1.000000        1.000000      1.000000   
75%         2.000000         1.000000        2.000000      2.000000   
max         2.000000         2.000000        2.000000      2.000000   

       DeviceProtection  TechSupport  StreamingTV  StreamingMovies  \
count       7043.000000  7043.000000  7043.000000      7043.000000   
mean           0.904444     0.797104     0.985376         0.992475   
std            0.879949     0.861551     0.885002         0.885091   
min            0.000000     0.000000     0.000000         0.000000   
25%            0.000000     0.000000     0.000000         0.000000   
50%            1.000000     1.000000     1.000000         1.000000   
75%            2.000000     2.000000     2.000000         2.000000   
max            2.000000     2.000000     2.000000         2.000000   

          Contract  PaperlessBilling  PaymentMethod  MonthlyCharges  \
count  7043.000000       7043.000000    7043.000000     7043.000000   
mean      0.690473          0.592219       1.574329       64.761692   
std       0.833755          0.491457       1.068104       30.090047   
min       0.000000          0.000000       0.000000       18.250000   
25%       0.000000          0.000000       1.000000       35.500000   
50%       0.000000          1.000000       2.000000       70.350000   
75%       1.000000          1.000000       2.000000       89.850000   
max       2.000000          1.000000       3.000000      118.750000   

       TotalCharges        Churn  
count   7032.000000  7043.000000  
mean    2283.300441     0.265370  
std     2266.771362     0.441561  
min       18.800000     0.000000  
25%      401.450000     0.000000  
50%     1397.475000     0.000000  
75%     3794.737500     1.000000  
max     8684.800000     1.000000

SeniorCitizen is actually a categorical hence the 25%-50%-75% distribution is not propoer

75% customers have tenure less than 55 months

Average Monthly charges are USD 64.76 whereas 25% customers pay more than USD 89.85 per month

In [33]:
predict = "Churn"
data1 = data[["gender","SeniorCitizen","Partner","Dependents","PhoneService","MultipleLines","InternetService","OnlineSecurity","OnlineBackup","DeviceProtection","TechSupport","StreamingTV","StreamingMovies","Contract","PaperlessBilling","PaymentMethod","MonthlyCharges","TotalCharges","Churn","Tenure"]]

In [34]:
x = np.array(data1.drop([predict], axis=1))
y = np.array(data1[predict])

In [35]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2)

In [36]:
nan_rows = np.isnan(x_train).any(axis=1)
# Remove rows with missing values from both x_train and y_train
x_train = x_train[~nan_rows]
y_train = y_train[~nan_rows]

In [37]:
nan_rows = np.isnan(x_test).any(axis=1)

# Remove rows with missing values from both x_test and y_test
x_test = x_test[~nan_rows]
y_test = y_test[~nan_rows]

In [38]:
#Training Different Classification models
xgb_model = XGBClassifier().fit(x_train,y_train)
lgbm_model = LGBMClassifier().fit(x_train,y_train)
dt_model = DecisionTreeClassifier().fit(x_train,y_train)
hgb_model = HistGradientBoostingClassifier().fit(x_train,y_train)
RF_model = RandomForestClassifier().fit(x_train,y_train)
gb_model = GradientBoostingClassifier().fit(x_train,y_train)
svc_model = svm.SVC().fit(x_train,y_train)

[LightGBM] [Info] Number of positive: 1493, number of negative: 4131
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000304 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 559
[LightGBM] [Info] Number of data points in the train set: 5624, number of used features: 19
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.265469 -> initscore=-1.017732
[LightGBM] [Info] Start training from score -1.017732


In [39]:
#creating a function to test the model and display its metrics
def evaluate_model(model,x_test,y_test):
    prediction = model.predict(x_test)
    print("Mean Absolute error", mean_absolute_error(y_test,prediction))
    print("Mean Absolute Percentage error", mean_absolute_percentage_error(y_test,prediction))
    print("Mean Squared error", mean_squared_error(y_test,prediction))
    print("Root Mean Squared error", np.sqrt(mean_squared_error(y_test,prediction)))
    print("R2 Score", r2_score(y_test,prediction))
    print("accuracy score:",accuracy_score(y_test,prediction))

In [40]:
#printing the values of the metrics of all the models
print("model: XG Boost Classifier")
print(evaluate_model(xgb_model,x_test,y_test))
print()
print("model: LGBM Classifier")
print(evaluate_model(lgbm_model,x_test,y_test))
print()
print("model: Decision Tree Classifier")
print(evaluate_model(dt_model,x_test,y_test))
print()
print("model: Hist Gradient Boosting Classifier")
print(evaluate_model(hgb_model,x_test,y_test))
print()
print("model: Random Forest Classifier")
print(evaluate_model(RF_model,x_test,y_test))
print()
print("model: Gradient Boosting Classifier")
print(evaluate_model(gb_model,x_test,y_test))
print()
print("model: SVM Classifier")
print(evaluate_model(svc_model,x_test,y_test))

model: XG Boost Classifier
Mean Absolute error 0.22088068181818182
Mean Absolute Percentage error 409418147942772.56
Mean Squared error 0.22088068181818182
Root Mean Squared error 0.46997944829341404
R2 Score -0.128484248721755
accuracy score: 0.7791193181818182
None

model: LGBM Classifier
Mean Absolute error 0.2109375
Mean Absolute Percentage error 364638038011531.75
Mean Squared error 0.2109375
Root Mean Squared error 0.4592793267718459
R2 Score -0.07768431469569537
accuracy score: 0.7890625
None

model: Decision Tree Classifier
Mean Absolute error 0.28551136363636365
Mean Absolute Percentage error 671701648968610.9
Mean Squared error 0.28551136363636365
Root Mean Squared error 0.5343326338867612
R2 Score -0.45868381989114315
accuracy score: 0.7144886363636364
None

model: Hist Gradient Boosting Classifier
Mean Absolute error 0.21875
Mean Absolute Percentage error 403020989381166.6
Mean Squared error 0.21875
Root Mean Squared error 0.46770717334674267
R2 Score -0.11759854857331375
a

In [41]:
#saving the model using pickle and choosing the Gradient Boosting classifier algorithm
import pickle
pickle.dump(gb_model,open("model.pkl","wb"))